Source/WebKit/NetworkProcess/NetworkSession.cpp

```diff
@@ -714,6 +715,48 @@ void NetworkSession::setEmulatedConditions(std::optional<int64_t>&& bytesPerSeco
 
 #endif // ENABLE(INSPECTOR_NETWORK_THROTTLING)
 
+static double connectionTimesMovingAverage(const Deque<Seconds, 25>& connectionTimes)
+{
+    constexpr double alphaSmoothing { 0.75 };
+    // EWMA:
+    // s_0 = x_0
+    // s_t = a * x_{t-1} + (1 - a) * s_{t-1}
+    //
+    // Where x_0 is the average of all recent connection timings, and alpha is 0.75.
+
+    double average = std::accumulate(connectionTimes.begin(), connectionTimes.end(), Seconds { }).seconds() / connectionTimes.size();
+    std::for_each(connectionTimes.begin(), connectionTimes.end(), [&average, alphaSmoothing](auto&& timing) {
+        average = alphaSmoothing*timing.seconds() + (1 - alphaSmoothing)*average;
+    });
+
+    return average;
+}
+
+void NetworkSession::recordHTTPSConnectionTiming(const NetworkLoadMetrics& metrics)
+{
+    constexpr double minumumConnectionTimeout { 3 };
+    constexpr double computedTimeoutScalingFactor { 1.5 };
+
+    if (metrics.reusedConnection())
+        return;
+
+    Seconds connectionEstablishmentTime = metrics.connectEnd - metrics.secureConnectionStart;
+    if (connectionEstablishmentTime.seconds() <= 0.0)
+        return;
+
+    auto& recentTiming = m_recentHTTPSConnectionTiming;
+    if (recentTiming.recentConnectionTimings.size() >= recentTiming.maxEntries)
+        recentTiming.recentConnectionTimings.removeFirst();
+    recentTiming.recentConnectionTimings.append(connectionEstablishmentTime);
+
+    auto newMovingAverage = connectionTimesMovingAverage(recentTiming.recentConnectionTimings);
+    newMovingAverage = std::max(minumumConnectionTimeout, newMovingAverage*computedTimeoutScalingFactor);
+    if (newMovingAverage != recentTiming.currentMovingAverage) {
```

Comment:
These will basically never be equal in practice.

Reply:
This will happen when we're repeatedly below the minimum timeout (3 seconds).